### PR TITLE
add example of using import maps for extensionless imports

### PR DESCRIPTION
**Pull request by devsnek**

### CODE REVIEW

**Comment by MylesBorins** [comment]
while I agree that import maps can do this, I don't feel like this is a scalable approach as it would require an entry for every file in every module that wants to use exetnsionless importing

**Comment by jkrems** [comment]
It would also work exclusively work bare specifiers (which is neat!) but doesn't really address what we have been calling "extension-less imports" in other contexts since those discussion were mostly about relative imports. I think the current proposal rightfully doesn't prescribe that `lodash/map` *needs* to be a file that is literally in the root of the lodash package, allowing for better internal organization of packages.

**Comment by devsnek** [comment]
@MylesBorins hopefully this doesn't come across as a recommendation, just as a demonstration of how certain patterns from different ecosystems can be recreated with this format. i agree that in the common case this will create very large files.

**Comment by jkrems** [comment]
Hah, guess I didn't follow this proposal closely enough. Sorry, got it wrong! `import:` URLs did bring the ability to remap relative imports. It's just somewhat buried in the README. So I assume a working lodash example would be, assuming that lodash itself uses relative extension-less imports in `lodash.js`:

```
{
  "imports": {
    "lodash": "/node_modules/lodash-es/lodash.js",
    "lodash/": "/node_modules/lodash-es/",
    "lodash/map": "/node_modules/lodash-es/map.js"
  }
  "scopes": {
    "/node_modules/lodash-es": {
      "./map": "/node_modules/lodash-es/map.js"
    },
    "/node_modules/lodash-es/fp": {
      "../map": "/node_modules/lodash-es/map.js"
    }
  }
}
```

That feels like it would become a mouthful for a module with more than a few files and any kind of directory nesting but it would technically be possible.

File: README.md [hunk 1]
```diff
@@ -190,7 +190,21 @@ import localeData from "moment/locale/zh-cn.js";
...
+#### Extension-less imports
+
+It is also common in the Node.js ecosystem to import files without including the extension. [we do not have the luxury](#the-nodejs-module-resolution-algorithm) of trying multiple file extensions until we find a good match. Fortunately, an import map can also include these cases.
+
+```diff
+ {
+   "imports": {
+     "lodash": "/node_modules/lodash-es/lodash.js",
+     "lodash/": "/node_modules/lodash-es/",
++    "lodash/fp": "/node_modules/lodash-es/fp.js",
```
Comment (zenparsing) [inline README.md]: This is an example of deep package imports, not really "extension-less imports".

In the context of the node modules WG discussions, extensionless imports refer to relative specifiers as well.

Comment (devsnek) [reply README.md]: it's kinda both, but i can see the confusion. do you think there is a better example?

**Comment by jkrems** [comment]
I think it's a bit misleading as-is because it suggests that all you'd need to do is adding one layer of import remapping and extension-less imports would be a reality. But for a package like `react-dom`, and assuming that we wouldn't want to parse and analyze every single file ahead of time, it would be >800 entries for this one package alone (!).

I think there's a difference here between "this is theoretically possible" and "this is practical".

**Comment by jkrems** [comment]
Btw, the number above doesn't even take into account things like directory searching etc. or the problem with predicting extension precedent.

**Comment by ljharb** [comment]
I'm not sure that assumption holds; I don't see any issue with parsing every single file ahead of time.

If the size of the import map becomes a problem, then that will be social pressure for package authors not to use extensionless paths. There's no need to forbid it at the platform level (or in the case of this PR, at the "documented examples" level)

**Comment by jkrems** [comment]
> If the size of the import map becomes a problem, then that will be social pressure for package authors not to use extensionless paths. There's no need to forbid it at the platform level (or in the case of this PR, at the "documented examples" level)

This reads to me like "let's just have people run into the open blade so they learn their lesson". And then we'd spend the next decade wading through an ecosystem full of packages that come with a warning label ("this looks like it should work in the browser but won't unless you use tool X first"). Maybe I'm misunderstanding something about what you're saying?

**Comment by ljharb** [comment]
@jkrems I don't anticipate it will be a problem in practice, because I personally think that everyone will continue to use a bundler, which will handle this for them. The only people who will run into trouble are those who want to omit a build process entirely, and I think it's appropriate for that desire to prove itself out as something the community actually wants as a group.

**Comment by guybedford** [comment]
> @jkrems I don't anticipate it will be a problem in practice, because I personally think that everyone will continue to use a bundler

Yes but bundlers will start adapting to shared library outputs with package maps.

Remember as well that the package map must load before ANY module JS can be fetched. As a result a small increase in the size of the package map will have a significant effect on load time.

**Comment by ljharb** [comment]
@guybedford Since most applications use a great many number of bare imports, and deep imports within those, that seems to be an already-present problem with import maps - perhaps it should be addressed separately?

**Comment by guybedford** [comment]
It’s a very succinct output when we just have two entries per package (main
and sub path trailing slash map), coupled with the multi version overrides
in scopes.

The issue is having to explicitly declare extension maps in app their
variations both internal and external to the package.

**Comment by jkrems** [comment]
To add some color to the number I mentioned above: import maps completely opt out of usual path resolution and a single-step only. Which means that for every directory-to-directory transition, a different import map entry is needed to support extension-less imports widely. So in the worst case every single import statement needs its own custom import map entry. Example:

```bash
/node_modules/my-pkg
  # export { someHelper } from './src/util';
  # export { SomeClass } from './src/class';
  /interface.mjs
  /src
    # import { someHelper } from './util'; export class SomeClass {}
    /class.mjs
    # export const someHelper = 42;
    /util.mjs
  /test
    # import { someHelper} from '../src/util';
    /util.test.mjs
```

This project has 4 imports of util. Every single one of them needs their own unique entry in the import map. Because the number of entries in the import map generally is `(#directories + 1) * #files`. And again, that's just counting adding the extension. It doesn't cover directory searching yet. Once you *also* include support for directories with `package.json` in random relative imports, it starts to become really hard to imagine that this would be a practical thing to support, even in a tool generating a map.

I'll also revise my "unless you parse every file and generate the map". That won't be possible because dynamic import and eval are a thing. So you'll have to generate the full set of all possible imports by combining every single file with every single referrer and enumerate every path that could lead from referrer to target.

**Comment by devsnek** [comment]
this isn't about figuring out if doing this makes sense for your application or not, it's about adding a more complete example set of import map's technical capacity. if you want to have a discussion about whether you should use this in your application please do so in another location.

**Comment by guybedford** [comment]
Modules group, meet import maps :)

**Comment by domenic** [comment]
Thanks all for the spirited discussion. One important thing that this highlighted is that not everyone is aware that import maps will change the behavior of relative URL-like specifiers as well, so I will want to add an example to specifically highlight that.

I think this particular example is worth keeping, because as @devsnek says, it is a good demonstration of the technical capabilities. But I will add some caveats and gentle warnings away from this pattern, since there is evidently debate about whether it should be encouraged. (And I personally side with those that discourage it.)

File: README.md [hunk 2]
```diff
@@ -190,7 +191,21 @@ import localeData from "moment/locale/zh-cn.js";
...
+ {
+   "imports": {
+     "lodash": "/node_modules/lodash-es/lodash.js",
+     "lodash/": "/node_modules/lodash-es/",
+     "lodash/fp": "/node_modules/lodash-es/fp.js",
+   }
+ }
+```
+
+would allow not only `import fp from "lodash/fp.js"`, but also allow `import fp from "loadsh/fp"`. Also note that it enable `import fp from "./fp"` from inside the `/node_modules/lodash-es/lodash.js` file or other siblings.
```
Comment (guybedford) [inline README.md]: > Also note that it enable `import fp from "./fp"` from inside the `/node_modules/lodash-es/lodash.js` file or other siblings.

Typo: enable**s**

I'm not sure this is true for the above configuration without another explicit entry in the map for this?

Comment (domenic) [reply README.md]: Oh, you're right, drat. I'm going to force-push.

**Comment by domenic** [comment]
I've posted a caveats PR at https://github.com/WICG/import-maps/pull/111 which I'd love feedback on from the others in this thread.

**Pull request merged**